cmake: Revamp handling of data files

hebasto opened the pull request:
This PR leverages the approach from the https://github.com/chaincodelabs/libmultiprocess project and introduces a new functions `target_json_data_sources()` and `target_raw_data_sources()`, which minimize the amount of code required to assign to assign a `*.json` or `*.raw` data file to the `test_bitcoin`,  `bench_bitcoin` or `unitester` targets.

As requested in https://github.com/bitcoin/bitcoin/pull/30901#issuecomment-2654622689, the `codegen` build target is now supported, if available:
```
$ cmake --version
cmake version 3.31.5

CMake suite maintained and supported by Kitware (kitware.com/cmake).
$ cmake -G "Ninja" -B build
$ cmake --build build --target codegen
```

hebasto commented:
Friendly ping @ryanofsky :)

hebasto commented on src/test/CMakeLists.txt:
```diff
@@ -144,6 +122,23 @@ add_executable(test_bitcoin
+    data/sighash.json
+    data/tx_invalid.json
+    data/tx_valid.json
+  RAW_FILES
+    data/asmap.raw
+  RAW_NAMESPACE test::data
```
Once such a scenario happens, we can use another call:
```cmake
target_data_sources(test_bitcoin
  RAW_FILES
    data/some_file_with_another_namespace.raw
  RAW_NAMESPACE somewhere::another_namespace
)
```
And I think it is better than (1) specifying `data/some_file_with_another_namespace.raw` to generate a header than (2) adding `${CMAKE_CURRENT_BINARY_DIR}/data/some_file_with_another_namespac.raw.h` to the sources.

fjahr commented:
tACK 0c35be69cf2a18a7a9173d0f9f88116b4417c892

I have based #28792 on this now. It works and seems like a nicer way to do what I need there than the previous approach.

Take it with a grain of salt though because I don't think I am experienced enough with CMake to tell if there are any hidden issues with this approach or if there are even better ways to achieve this.

hebasto commented:
Rebased due to the conflict with the merged bitcoin/bitcoin#31503.

hebasto commented:
Rebased due to the conflict with the merged bitcoin/bitcoin#31542.

Sjors commented:
ACK 4185ad1cb12060337c2d66a9a66e37be1d10f4ce

My cmake knowledge is also limited, but I like that this removes duplication. E.g instead of `base58_encode_decode.json.h` and `data/base58_encode_decode.json` now `test/CMakeLists.txt` only contains the latter.

The build log looks the same, before and after:

```
[ 76%] Generating data/base58_encode_decode.json.h
```

hebasto commented:
Friendly ping @theuni :)

theuni commented on cmake/module/TargetDataSources.cmake:
```diff
@@ -0,0 +1,36 @@
+# Distributed under the MIT software license, see the accompanying
+# file COPYING or https://opensource.org/license/mit/.
+
+function(target_data_sources target)
```
I think this would make sense as 2 separate functions. Seems a weird mix of args otherwise.
`target_raw_data_source` and `target_json_data_source`?

I think it'd be fine to call both funcs for targets that need both types of files generated.

hebasto replied:
Thanks! [Reworked](https://github.com/bitcoin/bitcoin/pull/30901#issuecomment-2656487136).

theuni commented on cmake/module/TargetDataSources.cmake:
```diff
@@ -0,0 +1,36 @@
+      COMMAND ${CMAKE_COMMAND} -DJSON_SOURCE_PATH=${CMAKE_CURRENT_SOURCE_DIR}/${json_file} -DHEADER_PATH=${CMAKE_CURRENT_BINARY_DIR}/${json_file}.h -P ${PROJECT_SOURCE_DIR}/cmake/script/GenerateHeaderFromJson.cmake
+      DEPENDS ${CMAKE_CURRENT_SOURCE_DIR}/${json_file} ${PROJECT_SOURCE_DIR}/cmake/script/GenerateHeaderFromJson.cmake
+      MAIN_DEPENDENCY ${CMAKE_CURRENT_SOURCE_DIR}/${json_file}
+      VERBATIM
```
These are now missing `DEPENDS_EXPLICIT_OPT` from #30911.

hebasto replied:
Thanks! [Rebased](https://github.com/bitcoin/bitcoin/pull/30901#issuecomment-2656487136).

theuni reviewed (commented):
Concept ACK.

theuni commented:
Also, since you're messing with this, please consider the `CODEGEN` option for `add_custom_command`, which would need feature-gating same as `DEPENDS_EXPLICIT_OPT`.

hebasto committed:
cmake: Add support for builtin `codegen` target

Additionally, this change removes unnecessary braces in the `if()`
command for improved robustness, readability and consistency with CMake
guidelines.

hebasto commented:
1. Rebased due to the conflict with the merged bitcoin/bitcoin#30911.
2. Addressed the recent @theuni's feedback.
3. Fixed minor bugs, such as: https://github.com/bitcoin/bitcoin/blob/18cc0a55595f9dc1f2e561743201a05754996b64/cmake/module/TargetDataSources.cmake#L6

> Also, since you're messing with this, please consider the `CODEGEN` option for `add_custom_command`, which would need feature-gating same as `DEPENDS_EXPLICIT_OPT`.

Nice! Added.

Please note that `cmake_policy(SET CMP0171 NEW)` within `TargetDataSources.cmake` does not work for some reason. However, this behaviour is not documented by CMake.

fjahr commented:
tACK 18cc0a55595f9dc1f2e561743201a05754996b64

Again, I am not an expert on cmake but I don't see any downsides in the latest changes and I can confirm it works. I am using this latest version in #28792 now. And I find the change on function names and args is making things a bit nicer.

hebasto commented:
> Hmm, something here has broken the dependency flattening. I'm seeing a stall again due to the file generation.

What CMake version?

hebasto committed:
cmake: Revamp handling of data files

This change introduces new functions `target_json_data_sources()` and
`target_raw_data_sources()`.

hebasto commented:
@theuni 

> Hmm, something here has broken the dependency flattening. I'm seeing a stall again due to the file generation.

I've removed the `MAIN_DEPENDENCY` options (I don't recall the motivation behind their introduction). A comparison of the resulting `build.ninja` files reveals no changes in dependencies.

theuni commented:
> I've removed the `MAIN_DEPENDENCY` options (I don't recall the motivation behind their introduction). A comparison of the resulting `build.ninja` files reveals no changes in dependencies.

I believe `MAIN_DEPENDENCY` is MSVC related, so it wouldn't show up in the `build.ninja`. Did it help with something there?

theuni commented:
> @theuni
> 
> > Hmm, something here has broken the dependency flattening. I'm seeing a stall again due to the file generation.

Ok, Ignore this. I believe I was testing with make, where the stall is expected. Ninja is indeed unchanged.

theuni reviewed (approved):
ACK ecf54a32ed26a50e861fca559e43ec1f9dee93b7

fjahr commented:
re-ACK ecf54a32ed26a50e861fca559e43ec1f9dee93b7

Only removed `MAIN_DEPENDENCY` since last review.

Sjors commented:
re-tACK ecf54a32ed26a50e861fca559e43ec1f9dee93b7

The range-diff shows that we persevere the `DEPENDS_EXPLICIT_OPT` check added in #30911.

Checked the same thing as last time: https://github.com/bitcoin/bitcoin/pull/30901#issuecomment-2598371523 

> My cmake knowledge is also limited, but I like that this removes duplication.

It still is, and this PR still does.

hebasto commented:
> I believe `MAIN_DEPENDENCY` is MSVC related, so it wouldn't show up in the `build.ninja`. Did it help with something there?

[It](https://cmake.org/cmake/help/latest/command/add_custom_command.html):
> is treated just like any value given to the `DEPENDS` option but also suggests to Visual Studio generators where to hang the custom command.

I'm not sure how useful it is.
